teps/0076-array-result-types.md

```diff
@@ -0,0 +1,531 @@
+        |   or maybe the sharding is based on previous run information
+        | result: list of the names of tests each shard should run, e.g. [[test_foo, test_bar], [test_baz]]
+        v
+   (for each: divide-tests-into-shards result)
+      golang-test (the task would need to be updated to support running a subset of tests)
+   ```
+
+1. **Use case for array indexing: Unwinding loops**
+   Without looping support and without indexing, there will be no way to mix array results with tasks that expect
+   strings.
+
+   If we add support for array results before we add support for looping (and until we have an approve proposa, there is
+   no guarantee we ever will!) there will be no way to use an array result from one Task with a Task that expects a
+   string parameter.
+
+   For example, without looping, folks can get the reuse they want by hard coding the Tasks explicitly, which would
+   probably support a surprising number of use cases (but look ugly and involve a lot of copy and paste), e.g.:
+
+  * Have a task that divides tests into 3 shards - if you know there are always 3 shards, you could add 3 go-test
+    pipeline tasks and have each one use the result of the task that divides them:
+    ```
+       divide-tests-into-shards
+                  |
+                  | result: list of starting and ending indexes for each shard, e.g. [[0, 7], [8, 14], [15, 23]]
+                / | \
+    go-test-0  go-test-1  go-test-2
+      [0,7]      [8,14]   [15, 23]
+    ```
+    
+  * If you know a repo contains 14 Dockerfiles (e.g.
+    [tekton/images in plumbing contains 14 dockerfiles](https://github.com/tektoncd/plumbing/tree/main/tekton/images) to
+    build, you could have a task that grabs the 14 paths, and then feed them into 14 kaniko pipeline tasks individually
+
+
+1. **Use case for array indexing: Tasks that wrap CLIs**
+   Array indexing would make it possible for Tasks that wrap generic CLIs to provide
+   (short - see [size limits](#size-limit)) stdout output as array results which could be indexed by consuming tasks.
+   For example [the gcloud CLI task](https://github.com/tektoncd/catalog/tree/main/task/gcloud/0.1): the CLI could be
+   doing all kinds of different things depending on the arguments (e.g. `gcloud projects list` vs
+   `gcloud container clusters list`). In order to consume the output in downstream tasks (with minimal effort), you
+   could index into the result. (Note that the existing gcloud task provides no results at all.)
+
+   Example gcloud output:
+
+   ```
+   # gcloud container clusters list
+   NAME        LOCATION     MASTER_VERSION   MASTER_IP     MACHINE_TYPE   NODE_VERSION         NUM_NODES  STATUS
+   eucalyptus  us-central1  1.18.20-gke.501  1.1.1.1  n1-standard-4  1.16.15-gke.6000 **  3          RUNNING
+   ```
+
+   Using this as a result would require which this TEP is not proposing but would be a logical next step (e.g.
```

Comment:
nit: missing word I think. `Using this as a result would require [something] which this TEP is not`

Reply:
whoops thanks @sbwsg !! i think [something] is 'supporting nested arrays'